const thread-safe packed classes

Dr15Jones opened the pull request:
Made the mutable members of pat::PackedCandidate and pat::PackedGenParticle being changed in const functions be thread-safe. Added unit tests.

Dr15Jones committed:
Made pat::PackedCandidate const thread-safe

Made the updating of mutable members in const functions of
pat::PackedCandidate thread-safe. Added a unit test for the class.

Dr15Jones committed:
Made pat::PackedGenParticle const thread-safe

Made the updating of mutable members in const functions of
pat::PackedGenParticle thread-safe. Added a unit test for the class.

Dr15Jones commented:
I didn't know of anyway to actually test that reading back the classes work correctly.

cmsbuild commented:
A new Pull Request was created by @Dr15Jones (Chris Jones) for CMSSW_7_6_X.

const thread-safe packed classes

It involves the following packages:

DataFormats/PatCandidates


@cmsbuild, @vadler, @monttj can you please review it and eventually sign? Thanks.
@gpetruc this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
JENKINS_TEST_URL

Dr15Jones commented on DataFormats/PatCandidates/interface/PackedGenParticle.h:
```diff
@@ -30,16 +34,65 @@ namespace pat {
+    : p4_(new PolarLorentzVector(c.pt(), c.eta(), c.phi(), c.mass())), p4c_( new LorentzVector(*p4_)), vertex_(0,0,0), pdgId_(c.pdgId()), charge_(c.charge()), mother_(c.motherRef(0)),
       statusFlags_(c.statusFlags()) { pack(); }
-  explicit PackedGenParticle( const reco::GenParticle & c, const edm::Ref<reco::GenParticleCollection> &  mother)
-    : p4_(c.pt(), c.eta(), c.phi(), c.mass()), p4c_(p4_), vertex_(0,0,0), pdgId_(c.pdgId()), charge_(c.charge()), mother_(mother), unpacked_(true),
```
NOTE: the vertex is not taken from the GenParticle. Is this intended?

cmsbuild commented:
-1 
Tested at: 9b8d11b6152466cbbcf92687eddd56563a3e1c7a
 I found an error when building: 
 
 <pre>>> Compiling edm plugin /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPhotonProducer.cc 
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc: In instantiation of 'pat::PATPackedCandidateProducer::sort_indexes(const std::vector<_RealType>&) const::<lambda(size_t, size_t)> [with T = pat::PackedCandidate; size_t = long unsigned int]':
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc:61:52:   required from 'struct pat::PATPackedCandidateProducer::sort_indexes(const std::vector<_RealType>&) const [with T = pat::PackedCandidate]::<lambda(size_t, size_t)>'
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc:61:117:   required from 'std::vector<long unsigned int> pat::PATPackedCandidateProducer::sort_indexes(const std::vector<_RealType>&) const [with T = pat::PackedCandidate]'
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc:291:52:   required from here
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc:61:116: error: no matching function for call to 'pat::PackedCandidate::PackedCandidate(const value_type&)'
               std::sort(idx.begin(), idx.end(),[&v,this](size_t i1, size_t i2) { return candsOrdering(v[i1],v[i2]);});
                                                                                                                    ^
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc:61:116: note: candidate is:
In file included from /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc:9:0:
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_7_6_X_2015-10-08-2300/src/DataFormats/PatCandidates/interface/PackedCandidate.h:34:5: note: pat::PackedCandidate::PackedCandidate()
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11706/8715/summary.html

Dr15Jones committed:
Removed incorrect explicit from copy and move constructors

Using explicit with copy and move constructors causes problems with
some common patterns. The version generated by default by the compiler
do not use explicit.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
JENKINS_TEST_URL

cmsbuild commented:
Pull request #11706 was updated. @cmsbuild, @vadler, @monttj can you please check and sign again.

cmsbuild commented:
+1
Tested at: b4e937a18c0fffeab7ebccd5cd02c45346ed7554
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11706/8725/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11706/8725/summary.html

wmtan commented:
@davidlange6 @Dr15Jones Chris asked me to look at this, but I have no idea how to reproduce the problem. I do not know which workflows reproduce the problem. nor how to generate and view the distributions that do not match.

Dr15Jones commented:
@wmtan I just drilled through the comparison to baseline plots and found 	50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50 to have a big difference if one clicks on that link, then goes to the 'ParticleFlow' and click on that pie chart then click on the pie chart for packedGenParticlesValidation then click on the 'CompWithGenParticles' pie chart.

So I suggest running workflow 50202.0. I searched from CompWithGenParticles and hit
http://cmslxr.fnal.gov/source/Validation/RecoParticleFlow/python/miniAODValidation_cff.py#0014
So you could look at that module to see what differences it is looking at.

wmtan commented:
@Dr15Jones The problem easily reproduces. In step3 of relval 50202.0, pat::PackedGenParticle is produced. The distributions of all data members (e.g. pt, Y, etc) are totally screwed up. The differences are far from subtle, and you can tell it is screwed up just by looking. No need to compare with the distributions prior to this PR to tell the distributions are wrong. I'll try to debug this tomorrow.
Some good news: The pat::PackedCandidate distributions look OK, so the problem is confined to
pat::PackedGenParticle.

davidlange6 commented:
and this may or may not be related - but lots of warnings coming (I think) from these changes

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc493/CMSSW_7_6_X_2015-10-19-1100/DataFormats/PatCandidates

> On Oct 20, 2015, at 1:27 AM, wmtan <notifications@github.com> wrote:
> 
> @Dr15Jones The problem easily reproduces. In step3 of relval 50202.0, pat::PackedGenParticle is produced. The distributions of all data members (e.g. pt, Y, etc) are totally screwed up. The differences are far from subtle, and you can tell it is screwed up just by looking. No need to compare with the distributions prior to this PR to tell the distributions are wrong. I'll try to debug this tomorrow.
> Some good news: The pat::PackedCandidate distributions look OK, so the problem is confined to
> pat::PackedGenParticle.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Dr15Jones commented:
We can make the warnings go away by using the macro CMS_THREAD_GUARD instead if the attribute directly.

wmtan commented:
#12000 fixes the discrepancies. However, I did #12000 before I saw the comment about the thread guard warnings, so it does not fix those. The warnings are unrelated to the discrepancy.
